app/backend/src/tests/test_auth.py

```diff
@@ -309,4 +309,272 @@ def test_signup_invalid_birthdate(db):
             assert e.value.details() == errors.INVALID_BIRTHDATE
 
 
+def test_invalid_signup_token_info(db):
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        with pytest.raises(grpc.RpcError) as e:
+            reply = auth_api.SignupTokenInfo(auth_pb2.SignupTokenInfoReq(signup_token="notarealtoken"))
+        assert e.value.code() == grpc.StatusCode.NOT_FOUND
+        assert e.value.details() == errors.INVALID_TOKEN
+
+
+def test_signup_invalid_email(db):
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b."))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email="a@b.c"))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.INVALID_EMAIL
+
+
+def test_signup_existing_email(db):
+    # Signed up user
+    user, _ = generate_user()
+
+    # Attempt to signup again with the same email
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Signup(auth_pb2.SignupReq(email=user.email))
+    assert reply.next_step == auth_pb2.SignupRes.SignupStep.EMAIL_EXISTS
+
+
+def test_successful_authenticate(db):
+    user, _ = generate_user()
+
+    # Authenticate with username
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Authenticate(auth_pb2.AuthReq(user=user.username, password="password"))
+    assert reply.jailed == False
+
+    # Authenticate with email
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        reply = auth_api.Authenticate(auth_pb2.AuthReq(user=user.email, password="password"))
+    assert reply.jailed == False
+
+
+def test_unsuccessful_authenticate(db):
+    user, _ = generate_user()
+
+    # Invalid password
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        with pytest.raises(grpc.RpcError) as e:
+            reply = auth_api.Authenticate(auth_pb2.AuthReq(user=user.username, password="incorrectpassword"))
+        assert e.value.code() == grpc.StatusCode.NOT_FOUND
+        assert e.value.details() == errors.INVALID_USERNAME_OR_PASSWORD
+
+    # Invalid username
+    with auth_api_session() as (auth_api, metadata_interceptor):
+        with pytest.raises(grpc.RpcError) as e:
+            reply = auth_api.Authenticate(auth_pb2.AuthReq(user="Notarealusername", password="password"))
```

Comment:
Please make username lowercase

Reply:
I saw the user id being an option here, but when I try passing that to user I get this error within auth_pb2.AuthReq
`TypeError: 1 has type int, but expected one of: bytes, unicode`

So I was thinking there was some type restriction on what could be passed to AuthReq but I couldn't figure out how to fix it. Do you have any suggestions, or do you even think it's important I test username, email, and user id?

Reply:
That field has to be passed in as a string, even if you're entering a number, due to the typing of gRPC.

But testing each of the three cases isn't *that* important, up to you if you feel like adding them.

Reply:
Just as you were responding I found the code that expected it to be a string. You say it's due to gRPC typing, would it be beneficial to add functionality that would accept an int as well? Or would that be just a convenience thing for testing and not actually used in production?

Reply:
We'd have to add another field, and then have something like 

```proto3
message WhateverItsCalled {
  oneof user {
    user_id = 1;
    user_username = 2;
    user_email = 3;
  }
}
```

it just moves the burden of regexping from the backend to the frontend. Easier to keep on the backend for now!

I think going forward we might want to get rid of the places where we do this though, user IDs should really never be exposed in the frontend

Reply:
One final question, is it prefered I add another commit to fix these changes, and then end up merging both of them into develop (if approved), or should I amend the current commit so only one gets merged into develop?

Reply:
Normally we add commits; it's more convenient for the reviewer to see the new changes, see the progression of CI/CD successes/failures, etc. But you can also amend if you prefer (as long as you only amend in a PR and not on `develop` as rewriting history is bad when multiple people have the branch checked out).